### PR TITLE
 Add Swagger and Redoc

### PR DESCRIPTION
**Pull request by Kukks**
Blocked by #1262

### CODE REVIEW

**Comment by Kukks** [comment]
@NicolasDorier  updated and rebased

**Comment by NicolasDorier** [comment]
Cool, where can I see the API doc?

**Comment by NicolasDorier** [comment]
Can you add one important feature? A Label. So people can remember what they issued their API key for,

**Comment by NicolasDorier** [comment]
Also, can you silently remove the `-` from API keys? So people can just double click to copy paste API keys.

**Comment by NicolasDorier** [comment]
The doc is really awesome.

That said, I try the endpoint you documented https://localhost:14142/api-keys/authorize

It crashed

```
fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware: An unhandled exception has occurred while executing the request.
System.ArgumentNullException: Value cannot be null. (Parameter 'source')
   at System.Linq.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
   at System.Linq.Enumerable.Contains[TSource](IEnumerable`1 source, TSource value, IEqualityComparer`1 comparer)
   at BTCPayServer.Controllers.ManageController.AuthorizeAPIKey(AuthorizeApiKeysViewModel viewModel) in C:\Sources\btcpayserver\BTCPayServer\Controllers\ManageController.APIKeys.cs:line 112
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
```

Also it does not show what permission the app is asking?

**Comment by Kukks** [comment]
> 
> 
> The doc is really awesome.
> 
> That said, I try the endpoint you documented https://localhost:14142/api-keys/authorize
> 
> It crashed
> 
> ```
> fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware: An unhandled exception has occurred while executing the request.
> System.ArgumentNullException: Value cannot be null. (Parameter 'source')
>    at System.Linq.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
>    at System.Linq.Enumerable.Contains[TSource](IEnumerable`1 source, TSource value, IEqualityComparer`1 comparer)
>    at BTCPayServer.Controllers.ManageController.AuthorizeAPIKey(AuthorizeApiKeysViewModel viewModel) in C:\Sources\btcpayserver\BTCPayServer\Controllers\ManageController.APIKeys.cs:line 112
>    at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
> ```
> 
> Also it does not show what permission the app is asking?

Fixed in next commit

`Also it does not show what permission the app is asking?`
it does show, when you specify permissions. Basically you can request an api key as a proof of account on a specific btcpayserver, without any permissions. (this is super awesome when we add back in the redirect feature)

File: BTCPayServer.Data/Migrations/20200225133433_AddApiKeyLabel.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace BTCPayServer.Migrations
+{
+    public partial class AddApiKeyLabel : Migration
```
Comment (NicolasDorier) [inline BTCPayServer.Data/Migrations/20200225133433_AddApiKeyLabel.cs]: can you put that in the other migration you did? On top of this, there is no attribute so this is probably skipped.

Comment (Kukks) [reply BTCPayServer.Data/Migrations/20200225133433_AddApiKeyLabel.cs]: done

**Comment by Kukks** [comment]
done @NicolasDorier

**Comment by NicolasDorier** [comment]
tests fail

**Comment by NicolasDorier** [comment]
Can you remove the `-` from API Keys so people don't bother us with copy/paste isssues?

**Comment by NicolasDorier** [comment]
image

**Comment by NicolasDorier** [comment]
Also, you need to improve the text on the `api-keys/authorize`.

image


What can it access ??

**Comment by Kukks** [comment]
> 
> 
> image

Are you sure you pulled latest?

**Comment by Kukks** [comment]
> 
> 
> Also, you need to improve the text on the `api-keys/authorize`.
> 
> image
> 
> What can it access ??
It only shows permissions that you request. If you do not request permissions, you do not get any shown there. 

The API key would not provide you any BTCPay functionality BUT you could use it as a way to do a "Login with BTCPay" feature on your application

**Comment by NicolasDorier** [comment]
@Kukks this is really feature creep, I thought this was for streamlining the api key creation.
The message is also not clear at all. You don't know what it mean to grant access tot he app

**Comment by Kukks** [comment]
It is, I can just add a message `There are no  associated permissions to the API key being requested here. The application cannot do anything with your BTCPay account other than validating your account exists.`

**Comment by NicolasDorier** [comment]
`docs/index.html?url=/swagger/v1/swagger.json` just keep spinning now.

Also can you remove the `-` from apikey?

**Comment by Kukks** [comment]
> 
> 
> `docs/index.html?url=/swagger/v1/swagger.json` just keep spinning now.
> 
> Also can you remove the `-` from apikey?

Already removed the `-` when new keys are generated.

Fixed the spinner

**Comment by Kukks** [comment]
added message too now

**Comment by NicolasDorier** [comment]
@Kukks fuck you need another migration for the label actually. The previous migration released in .161.

**Comment by NicolasDorier** [comment]
git pull because I fixed the colspan

**Comment by NicolasDorier** [comment]
You did not put an attribute on the migration, so I don't think this is picked up

**Comment by Kukks** [comment]
that is because the ef tooling generates a designer file with the attribute there, changed it now

**Comment by NicolasDorier** [comment]
A GET action should never have a non idempotent action.
The Remove link of the API should use the `Confirm` page as we do everywhere. (check how I did for ListStores for example)

**Comment by Kukks** [comment]
done @NicolasDorier

**Pull request merged**